source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc

```diff
@@ -33,6 +33,25 @@ DnsCacheImpl::DnsCacheImpl(
       host_ttl_(PROTOBUF_GET_MS_OR_DEFAULT(config, host_ttl, 300000)),
       max_hosts_(PROTOBUF_GET_WRAPPED_OR_DEFAULT(config, max_hosts, 1024)) {
   tls_slot_.set([&](Event::Dispatcher&) { return std::make_shared<ThreadLocalHostInfo>(*this); });
+
+  // Pre-Loaded hostnames are resolved without a read lock on primary hosts because it is done
+  // during object construction.
+  uint pre_loaded_hostnames{};
+  for (auto it = config.pre_load_hostnames().begin();
+       it != config.pre_load_hostnames().end() && pre_loaded_hostnames < max_hosts_;
```

Comment:
Could default `max_hosts_` here be too large? it would be a lot of dns queries in a short span of time, should a few proxies come up simultaneously.

Reply:
Interesting point. We could define a small (10?) limit in the proto. And do min of proto limit and max hosts in code.

Reply:
Wdyt @dmitri-d ?

Reply:
Perhaps to make changing the limit easier, instead of having a limit in the proto, set the limit in the code? Then you could leave it at 10, and we can adjust it if it becomes necessary...

Reply:
I feel a little iffy about enforcing a limit in the implementation that doesn't have visibility in the API.

On second thought, we have never limit the number of potential bootstrap (dns based) clusters that envoy comes up with, which would also issue dns resolutions to the main thread. So I am not sure that the performance bottleneck is something we explicitly have to protect against. Specially for an opt in feature in an extension. Further, if someone is preloading a bunch of hostnames into the dynamic forward proxy I would argue they might be better off branching their config to have dedicated clusters for those names.

Wdyt @dmitri-d?

Reply:
I don't feel strongly about this, tbh. My concern is very hypothetical and perhaps just pointing out a potential issue with large number of hosts in pre-load in docs is sufficient?

Reply:
Yep, I think that is sufficient. Will update docs. Thanks for the discussion.

Reply:
isn't max hosts also from config?  can we just reject config if you set max hosts to 5 and try to preload 10?

Reply:
I guess I could throw invalid config upon construction. I hadn't considered that. I am fine doing either (reject the config, or load up to max hosts).

@dmitri-d had concerns about pre-loading up to max hosts if max hosts is _too_ big.

Reply:
I'd prefer throwing, but I think it should at least warn that you're not doing what the config asked you to do.

Reply:
@alyssawilk do you have a preference?